python/triton_kernels/triton_kernels/numerics_details/mxfp_details/_downcast_to_mxfp.py

```diff
@@ -165,7 +200,43 @@ def _downcast_to_mxfp(mx_tensor_ptr, stride_mxt_outer, stride_mxt_quant: tl.cons
                                                         DEQUANT_SCALE_ROUNDING_MODE)
 
     tl.store(mx_scale_ptr + mx_scale_offsets, scale_tensor, mask=full_scale_mask)
-    tl.store(mx_tensor_ptr + mx_tensor_offsets, out_tensor, mask=full_mask_mxt)
+    if is_fp4:
+        # Enabled grouped stores for FP4 elements.
+        GROUP_BYTES: tl.constexpr = 4
+        GROUPS_PER_ROW: tl.constexpr = BLOCK_SIZE_QUANT_MX_TENSOR // GROUP_BYTES
+
+        # Compute group-aligned address offsets for 4-byte stores.
+        offs_mxt_quant_groups = (tl.arange(0, GROUPS_PER_ROW)[None, :].to(tl.int64) * GROUP_BYTES)
+        mx_tensor_group_offsets = offs_mxt_quant_groups * stride_mxt_quant + offs_outer * stride_mxt_outer
+
+        # Determine which groups are fully valid (ie, have 4 bytes in-bounds for this tile).
+        bound_elems = tl.cdiv(quant_dim, K_DIVISOR)
+        groups_last_idx = start_mx_quant + offs_mxt_quant_groups + (GROUP_BYTES - 1)
+        mask_groups = groups_last_idx < bound_elems
+        full_group_mask = mask_groups & mask_n
+
+        # Only 4-byte store on rows whose base address is 4-byte aligned.
+        row_base = start_mx_quant + offs_outer * stride_mxt_outer
+        aligned_rows = (row_base & (GROUP_BYTES - 1)) == 0
+        full_group_mask = full_group_mask & aligned_rows
+
+        # Pack 4 consecutive bytes into one u32.
+        vals_u32 = out_tensor.to(tl.uint32)
+        vals4 = tl.reshape(vals_u32, [BLOCK_SIZE_OUT_DIM, GROUPS_PER_ROW, GROUP_BYTES])
+        shift_amounts = (tl.arange(0, GROUP_BYTES)[None, None, :].to(tl.uint32) * 8)
+        packed_u32 = tl.sum(vals4 << shift_amounts, axis=2)
+
+        mx_tensor_u32_ptr = (mx_tensor_ptr + mx_tensor_group_offsets).to(tl.pointer_type(tl.uint32))
+        tl.store(mx_tensor_u32_ptr, packed_u32, mask=full_group_mask)
```

Comment:
You need to check if the `mx_tensor_ptr` is a multiple of 4, because if it isn't, you're doing unaligned stores, which is a UB